feat: include local dirty changes by default. fixes #184

sabard opened the pull request:
A quick attempt at implementing the changes described in [this comment](https://github.com/copier-org/copier/issues/184#issuecomment-861639395).
 - if a template is vcs-tracked and dirty, the changes are committed automatically with a "wip" commit and propagated to the subproject
 - user is warned in this scenario
 - tests for copy and update on dirty templates

I made this the default behavior, but would be happy to add a flag for this behavior and maintain the current default.

Also, think this will have to rebase off of #519 to pass CI

pawamoy commented on copier/template.py:
```diff
@@ -426,11 +427,20 @@ def templates_suffix(self) -> str:
+            if result.is_dir():
+                with local.cwd(result):
+                    if bool(git("status", "--porcelain").strip()):
+                        git("add", ".")
```
Do we want `-A` as well? Without it, removed files would not be added, leaving some dirtiness.

sabard replied:
Yep, good catch. Although removed files are currently not staged anyway during update—looks like the same issue as in #461. I took a stab fixing it by removing files that were present in `dst` and not `src`, but this is difficult when you factor in templating and it was causing errors in other tests.

pawamoy commented on copier/template.py:
```diff
@@ -426,11 +427,20 @@ def templates_suffix(self) -> str:
+                with local.cwd(result):
+                    if bool(git("status", "--porcelain").strip()):
+                        git("add", ".")
+                        git("commit", "-am", "wip")
```
If everything was added by `git add`, `-a` is not necessary.

pawamoy commented on tests/test_dirty_local.py:
```diff
@@ -0,0 +1,85 @@
+    # dst must be vcs-tracked to use run_update
+    with local.cwd(dst):
+        git("init")
+        git("add", ".")
+        git("commit", "-am", "first commit on dst")
```
Same two remarks as above (+`-A` and -`-a`)

pawamoy commented:
So, just to be clear: this change creates the wip commits in the temporary copy/clone, right? Not in the original template repository?

sabard commented:
I originally was creating the commit in the original template repository because that was easier, but I agree that it is too intrusive. 

The current code modifies `vcs::clone` instead of `template::local_abspath` and copies a dirty repo into a temporary directory so that it can add, commit, and clone.

pawamoy commented on copier/vcs.py:
```diff
@@ -111,15 +114,36 @@ def checkout_latest_tag(local_repo: StrOrPath, use_prereleases: OptBool = False)
+        with local.cwd(url):
+            is_dirty = bool(git("status", "--porcelain").strip())
+        if is_dirty:
+            src = tempfile.mkdtemp(prefix=f"{__name__}.dirty.")
```
The temporary directory should probably be removed once we're done using it. Maybe by using a `with` statement and rearranging the code so that `_clone` is inside?

sabard replied:
great, fixed and moved the duplicate `_clone` logic to `_perform_clone`

pawamoy commented:
Docs failure is because mkdocs-autorefs 0.3.1 was installed, while mkdocstrings 0.17.0 was not. I should probably have tagged mkdocs-autorefs 0.4.0 to prevent that. Or better yet, upper bounds could be removed on Copier's dependencies 😛

sabard commented:
Got it—should I just add `mkdocs-autorefs = "^0.4.0"`  to `pyproject.toml`? I always use pip-tools and not too familiar with poetry.

pawamoy commented:
I wasn't clear: *mkdocstrings* should be updated (up to 0.17.0), not mkdocs-autorefs. But it should probably be done in another PR.

sabard commented:
I saw this was fixed in #518, so I just rebased instead.

pawamoy commented:
Oh yes, good catch.

Yajo commented on tests/test_dirty_local.py:
```diff
@@ -0,0 +1,107 @@
+    assert bool(git("status", "--porcelain").strip())
+
+
+# Will fail due to lingering deleted file until #461 is fixed
```
I use the keyword HACK to track... well, hacks 😄 

Also TODO to make clear what to do after the HACK is solved.

```suggestion
# Will fail due to lingering deleted file.
# HACK https://github.com/copier-org/copier/issues/461
# TODO Remove xfail decorator when fixed.
```

Yajo reviewed (changes requested):
Hi @sabard! Welcome to Copier 😊 

This PR is very nice already. There are some improvements to do below, but overall it's almost ready to merge.

Thank you very much for your contribution!

sabard commented on copier/main.py:
```diff
@@ -677,6 +676,7 @@ def run_update(self) -> None:
                         file=sys.stderr,
                     )
                     diff = diff_cmd("--inter-hunk-context=0")
+                print(diff)
```
yes, thanks for catching this!

sabard commented on copier/vcs.py:
```diff
@@ -108,23 +111,61 @@ def checkout_latest_tag(local_repo: StrOrPath, use_prereleases: OptBool = False)
+        git("checkout", ref or "HEAD")
+        git("submodule", "update", "--checkout", "--init", "--recursive", "--force")
+
+    return dst
```
think this is needed so that `clone` returns `location` as is used [here](https://github.com/copier-org/copier/blob/3fe1f42728a4d3ce40b564d63420764f43d98ef1/copier/template.py#L434). I could also change `clone` to call `_perform_clone` and separately return `location`—lmk what you prefer

sabard commented on README.md:
```diff
@@ -42,12 +42,12 @@ To create a template:
-  help: What is your project name?
+    type: str
+    help: What is your project name?
 
 module_name:
-  type: str
-  help: What is your Python module name?
+    type: str
+    help: What is your Python module name?
```
this was happening automatically from `poe lint`. will see if i can commit without it

sabard replied:
was only able to get this pushed out with a `--no-verify`. are there any local options for prettier i need to set to make this work?

sabard replied:
this started working again all of a sudden—not sure what changed

sabard commented:
Removed Python 3.7 incompatible `dirs_exist_ok` option in `shutil.copytree` and fixed a bug in `test_dirty_local.py` causing CI to error.

Should be ready for another CI run.

Yajo commented on tests/test_dirty_local.py:
```diff
@@ -0,0 +1,117 @@
+# Will fail due to lingering deleted file.
+# HACK https://github.com/copier-org/copier/issues/461
+# TODO Remove xfail decorator when fixed.
+@pytest.mark.xfail
```
```suggestion
@pytest.mark.xfail(strict=True)
```

This way we won't forget to do what the comment says.

Yajo commented:
Please also add a note in the changelog.

sabard commented on copier/vcs.py:
```diff
@@ -108,23 +112,73 @@ def checkout_latest_tag(local_repo: StrOrPath, use_prereleases: OptBool = False)
+                        if os.path.isdir(s):
+                            copytree(s, d)
+                        else:
+                            copy2(s, d)
```
This turned out to be not so simple for a few reasons:
- `git diff` does not, by default, give a diff of the untracked files (of interested here doing a dirty copy)
- `HEAD` might not exist (e.g., `git init` was just run)

That said, I was able to get something closer to what you are asking for working by using `git diff --no-index` and then excluding `*/.git/*` on the apply. This will of course copy over other unwanted files to the temporary clone directory, but they shouldn't make it into the subdirectory since they will be in the clone's .gitignore. I think this is arguably just as inefficient in the worst-case (large .gitignored file getting needlessly copied).

I tried getting something like ` git --git-dir=<clone_dst>/.git --work-tree=<template_src> diff HEAD` [SO source](https://stackoverflow.com/a/32840462/916969) to work to no avail. Seems that anytime `git-diff` is used without `--no-index`, it ignores untracked files.

Need to make a couple fixes for CI still, but pushed latest changes to avoid losing progress.

Yajo replied:
AFAIK `git stash` works both with staged and untracked files. Have you tried with it?

sabard commented:
The most recent commit should probably pass CI. However:

I'm seeing `error: git diff header lacks filename information when removing 1 leading pathname component` from the `test_worker_config_precedence` test (attempting to copy the copier template itself). This is only triggered when the working tree is dirty, of course, but would appreciate any help figuring out how to fix this. Should be able to reproduce the error by checking out the latest commit here, making a change, and running the tests.

sabard commented:
After some thinking, I figured it would be a better solution to create a new `.git` dir in a temporary location, but point it to the files in the template dir. This lets us add and commit the dirty (and untracked!) changes while avoiding the extra copy. Please see the latest code and let me know if you have any questions.

Yajo commented on CHANGELOG.md:
```diff
@@ -31,6 +31,7 @@ Summary:
 -   Raise a CopierAnswersInterrupt instead of a bare KeyboardInterrupt to provide
     callers with additional context - such as the partially completed AnswersMap.
 -   Support for `user_defaults`, which take precedence over template defaults.
+-   Copies dirty changes from template to subdirectory by default.
```
```suggestion
-   Copy dirty changes from a git-tracked template to the project by default, to make testing easier.
```

Yajo commented on tests/test_dirty_local.py:
```diff
@@ -0,0 +1,117 @@
+        copier.copy(str(src), str(dst), data=DATA, quiet=True)
+
+    generated = (dst / "pyproject.toml").read_text()
+    control = (Path(__file__).parent / "reference_files/pyproject.toml").read_text()
```
```suggestion
    control = (Path(__file__).parent / "reference_files" / "pyproject.toml").read_text()
```

Yajo commented on copier/vcs.py:
```diff
@@ -108,23 +111,68 @@ def checkout_latest_tag(local_repo: StrOrPath, use_prereleases: OptBool = False)
+                with local.cwd(src):
+                    git("--git-dir=.git", f"--work-tree={url_abspath}", "init")
+                    git("--git-dir=.git", f"--work-tree={url_abspath}", "add", "-A")
+                    git(
```
This is a smart solution!

Have you tried this?:

1. Clone.
2. Do this same hack, but on the cloned directory.

sabard replied:
This is possible, but cloning a dirty git-tracked template seems more error prone to me. For example, it may not have a HEAD. I believe the submodule update would also have to be avoided and then called after the hack.

Yajo replied:
Ok, but in that case I still have some doubts about if this will work fine for updates, because if I'm understanding the code correctly, this is not copying the git history correctly, so it wouldn't work as expected for updates. So please could you update the test? (See below)

Note: git has no problem in cloning empty repos AFAIK.

sabard replied:
yeah, you're totally right. had to move things around a little bit to account for submodules, but this looks to be working now after cloning first

Yajo commented on tests/test_dirty_local.py:
```diff
@@ -0,0 +1,117 @@
+                Lorem ipsum
+            """,
+            src
+            / "to_delete.txt": """
```
Since the failure from this test is in this file, then please remove this part of the test and remove the xfail decorator. If the algorithm is fine, the test will pass. Later, when #461 is fixed, it will include the needed tests.

sabard replied:
kept file, but removed the assert. added a node referencing the linked issue

Yajo commented on copier/vcs.py:
```diff
@@ -150,29 +126,36 @@ def clone(url: str, ref: OptStr = None) -> str:
-
-    return _perform_clone(url, location, ref)
+            with local.cwd(location):
+                # git("--git-dir=.git", f"--work-tree={url_abspath}", "init")
```
Please remove this line.

Yajo commented on copier/vcs.py:
```diff
@@ -150,29 +126,36 @@ def clone(url: str, ref: OptStr = None) -> str:
+
+    with local.cwd(location):
+        git("checkout", ref or "HEAD")
+        git("submodule", "update", "--checkout", "--init", "--recursive", "--force")
```
I've got the feeling that this block should be done before importing the ditty changes. However, if the test goes green I guess I'm just wrong.

sabard replied:
I think it shouldn't matter if it runs before or after. But if there is no HEAD in the case of a recently intialized repo, then this fails and would need to be called after as well. I also realized that specifying `ref` means that dirty changes won't be considered and so added that to the dirty checking logic.

Yajo reviewed (commented):
Looks much nicer now!

Just remove that leftover line and one question to merge!

Thanks again, great job!

sabard commented:
Any tips for figuring out what's wrong with CI on windows? I cloned and ran the tests on my windows box and they passed, so having trouble replicating to get more info on the error.

Yajo commented on tests/test_dirty_local.py:
```diff
@@ -0,0 +1,116 @@
+    p1 = str(src / "aaaa.txt")
+    p2 = str(dst / "aaaa.txt")
+    filecmp.clear_cache()
+    assert filecmp.cmp(p1, p2)
```
Convert these to Path and do: 

```suggestion
    assert p1.read_text() == p2.read_text()
```

It should fail equally, but the error will be more descriptive.

Yajo replied:
It seems you were hitting some weird scenario with fliecmp.cmp and it was ok.

Yajo reviewed (approved):
Looks good! Just make it green to merge. 

Thanks.
